File: operators-example-java/src/main/java/net/bertag/operators/example/ReadingLog.java
package net.bertag.operators.example;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import net.bertag.operators.Ops;
import net.bertag.operators.api.Addable;
import net.bertag.operators.api.Scalable;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.function.BinaryOperator;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class ReadingLog implements Addable<ReadingLog>, Scalable<ReadingLog> {

    private static final BinaryOperator<LocalDate> minDate = BinaryOperator.minBy(Comparator.nullsLast(Comparator.naturalOrder()));
    private static final BinaryOperator<LocalDate> maxDate = BinaryOperator.maxBy(Comparator.nullsFirst(Comparator.naturalOrder()));

    private LocalDate startDate;
    private LocalDate endDate;
    private Integer pagesRead;
    private Double booksRead;

    public ReadingLog copy() {
        return new ReadingLog(startDate, endDate, pagesRead, booksRead);
    }

    @Override
    public ReadingLog plus(ReadingLog other) {
        startDate = minDate.apply(startDate, other.startDate);
        endDate = maxDate.apply(endDate, other.endDate);
        pagesRead = Ops.plus(pagesRead, other.pagesRead);
        booksRead = Ops.plus(booksRead, other.booksRead);
        return this;
    }

    @Override
    public ReadingLog scale(double factor) {
        pagesRead = Ops.scale(pagesRead, factor);
        booksRead = Ops.scale(booksRead, factor);
        return this;
    }
}

File: operators-example-kotlin/src/main/kotlin/net/bertag/operators/example/ReadingLog.kt
package net.bertag.operators.example;

import net.bertag.operators.api.Addable
import net.bertag.operators.api.Scalable
import net.bertag.operators.plus
import net.bertag.operators.scale
import java.time.LocalDate
import java.util.function.BinaryOperator

data class ReadingLog(
    var startDate: LocalDate?,
    var endDate: LocalDate?,
    var pagesRead: Int?,
    var booksRead: Double?
) : Addable<ReadingLog>, Scalable<ReadingLog> {

    private val minDate: BinaryOperator<LocalDate?> = BinaryOperator.minBy(Comparator.nullsLast(Comparator.naturalOrder()))
    private val maxDate: BinaryOperator<LocalDate?> = BinaryOperator.maxBy(Comparator.nullsFirst(Comparator.naturalOrder()))

    override fun plus(other: ReadingLog): ReadingLog {
        startDate = minDate.apply(startDate, other.startDate)
        endDate = maxDate.apply(endDate, other.endDate)
        pagesRead += other.pagesRead
        booksRead += other.booksRead
        return this;
    }

    override fun scale(factor: Double): ReadingLog {
        pagesRead = pagesRead.scale(factor)
        booksRead = booksRead.scale(factor)
        return this;
    }
}
